daml-lf/spec/daml-lf-1.rst

```diff
@@ -2569,6 +2575,150 @@ Map functions
+error.
+
+Of particular note, the following values are never valid keys:
+
+* Lambda expressions ``λ x : τ . e``
+* Type abstractions ``Λ α : k . e``
+* (Partially applied) built-in functions
+* (Un-evaluated) Update/Scenario values
+* Any value containing an invalid key
+
+**Comparison of Keys:** We define here the relation ``~ᵥ`` on value
+expressions that is used for key comparison. This is a partial
+equivalence relation over all values, but a (total) equivalence
+relation over valid keys.
+
+This relation is not exposed as a builtin function, but it coincides
+with the built-in equality for any given type, if that built-in isdefined. ::
+
+                                  ┌──────────┐
+  Generic Equivalence Relation    │ e₁ ~ᵥ e₂ │
+                                  └──────────┘
+
+  LitInt64₁ and LitInt64₂ have the same value
+  ——————————————————————————————————————————————————— GenEqLitInt64
+  LitInt64₁ ~ᵥ LitInt64₂
+
+  LitNumeric₁ and LitNumeric₂ have the same scale and value
+  ——————————————————————————————————————————————————— GenEqLitNumeric
+  LitNumeric₁ ~ᵥ LitNumeric₂
+
+  LitText₁ and LitText₂ have the same value
+  ——————————————————————————————————————————————————— GenEqLitText
+  LitText₁ ~ᵥ LitText₂
+
+  LitDate₁ and LitDate₂ have the same value
+  ——————————————————————————————————————————————————— GenEqLitDate
+  LitDate₁ ~ᵥ LitDate₂
+
+  LitTimestamp₁ and LitTimestamp₂ have the same value
+  ——————————————————————————————————————————————————— GenEqLitTimestamp
+  LitTimestamp₁ ~ᵥ LitTimestamp₂
+
+  cid₁ and cid₂ are the same
+  ——————————————————————————————————————————————————— GenEqLitContractId
+  cid₁ ~ᵥ cid₂
+
+  ——————————————————————————————————————————————————— GenEqUnit
+  () ~ᵥ ()
+
+  ——————————————————————————————————————————————————— GenEqTrue
+  'True' ~ᵥ 'True'
+
+  ——————————————————————————————————————————————————— GenEqFalse
+  'False' ~ᵥ 'False'
+
+  ——————————————————————————————————————————————————— GenEqListNil
+  'Nil' @τ₁ ~ᵥ 'Nil' @τ₂
+
+  e₁ ~ᵥ e₁'
+  e₂ ~ᵥ e₂'
+  ——————————————————————————————————————————————————— GenEqListCons
+  'Cons' @τ e₁ e₂  ~ᵥ 'Cons' @τ' e₁' e₂'
+
+  ——————————————————————————————————————————————————— GenEqOptionNone
+  'None' @τ₁ ~ᵥ 'None' @τ₂
+
+  e₁ ~ᵥ e₂
+  ——————————————————————————————————————————————————— GenEqOptionSome
+  'Some' @τ₁ e₁ ~ᵥ 'Some' @τ₂ e₂
+
+  e₁ ~ᵥ e₁'     …       eₙ ~ᵥ eₙ'
+  ——————————————————————————————————————————————————— GenEqRecCon
+  Mod:T @τ₁ … @τₙ { f₁ = e₁, …, fₙ = eₙ }
+    ~ᵥ Mod:T @τ₁' … @τₙ' { f₁ = e₁', …, fₙ = eₙ' }
+
+  e ~ᵥ e'
+  ——————————————————————————————————————————————————— GenEqVariantCon
+  Mod:T:V @τ₁ … @τₙ e ~ᵥ Mod:T:V @τ₁' … @τₙ' e'
+
+  ——————————————————————————————————————————————————— GenEqEnumCon
+  Mod:T:E ~ᵥ Mod:T:E
+
+  e₁ ~ᵥ e₁'     …       eₙ ~ᵥ eₙ'
+  ——————————————————————————————————————————————————— GenEqTupleCon
+  ⟨ f₁ = e₁, …, fₘ = eₘ ⟩ ~ᵥ ⟨ f₁ = e₁', …, fₘ = eₘ' ⟩
+
```

Comment:
You are forgetting the "nice" case where keys are maps. 
Of course we can a first implementation without map keys.

Reply:
That's not nice! We probably will need to add them, but I'm worried the "insertion order" issue will mess people up.

But also I don't see map values under the values section (including `'Map'`). Probably should add them?

Reply:
We can postpone this problem for later.

Reply:
TextMap (simply called 'Map' in the spec) are not concern with the insertion order.
They are sorted by key when converted to list or serialized in the ledger.

Reply:
Right. I propose that two `'Map'` (i.e. TextMap) are considered equal if they have the same key-value pairs, and two `'GenMap'` are considered equal if they have the same key-value pairs *in the same insertion order*, so they result in the same output when using `GENMAP_KEYS` and `GENMAP_VALUES`.

Reply:
(We could make the insertion order irrelevant for map key comparison, but then two "equal keys" could result in different outputs when passed through `GENMAP_KEYS`. Furthermore, the equality instance `Eq GenMap` that will be defined in DAML, in terms of `GENMAP_KEYS` and `GENMAP_VALUES`, is sensitive to insertion order.)

Reply:
Following discussion with @bame-da, we decided that equality for GenMaps should not consider insertion order.

Reply:
Are we planning to add a builtin `GENMAP_EQ : forall k v. (v -> v -> Bool) -> GenMap k v -> GenMap k v -> Bool`? (I assume we need to use the builtin equality for keys to make this work. I'm slowly starting to despise user-defined equality...)

Reply:
We were not planning to add this function as it was not here for `TextMap`. 

But I am planning to add a generic equality `EQUAL: forall v. v -> v -> Bool` using the builin equality. This will obviously work with `GenMap`. Is it not enough ?

Reply:
@remyhaemmerle-da The issue is that `Eq` (the DAML typeclass) does not necessarily match the built-in equality, so you may get unexpected results when using something like `EQUAL`. In any case we can define `GENMAP_EQ` in Daml without problem, it just won't be fast.

Reply:
@hurryabit Could we remove user-defined equality from DAML? Should we?

Reply:
OK. Let's start without it. We can still add it letter.

Reply:
I'm not averse to removing user-defined equality from DAML. Particularly if we find a way to keep the `Eq` type class as a marker but have no (or at least no user-implementable) methods in it. I'm not sure if that feasible though since `Eq` is probably wired quite deep into GHC.
@associahedron do you have some time at hand to investigate this a bit?
@bame-da what do you think about this?

Reply:
@hurryabit I think we can outlaw custom `Eq` instances in the preprocessor, and tell users to use `deriving Eq` instead.

Reply:
We could make it a warning for a while, and turn it into an outright error later.